File: tests/core/memop/pool.test.ts
import { Pool } from '../../../cocos/core/memop/pool';

class Test {
    public tag: number;
    public destroyed: boolean = false;
    constructor (num: number) {
        this.tag = num;
    }
}

let tagValue = 0;
function createTest () {
    return new Test(tagValue++);
}

let calledTimes = 0;
function destroyTest (obj) {
    calledTimes++;
    obj.destroyed = true;
}

const ARRAY_SIZE = 10;
let pool = new Pool<Test>(createTest, ARRAY_SIZE, destroyTest);

test('alloc', () => {
    for (let i = 0; i < ARRAY_SIZE + 2; ++i) {
        const obj = pool.alloc();
        expect(obj).not.toStrictEqual(undefined);
    }
});

test('free', () => {
    const tmp = createTest();
    pool.free(tmp);
    const newElement = pool.alloc();
    expect(newElement).toStrictEqual(tmp);
});

test('freeArray', () => {
    let tmpArr: Test[] = [];
    const TEST_SIZE = 5;
    for (let i = 0; i < TEST_SIZE; ++i) {
        tmpArr[i] = createTest();
    }
    pool.freeArray(tmpArr);
    for (let i = 0; i < TEST_SIZE; ++i) {
        const tmpObj = pool.alloc();
        expect(tmpArr[TEST_SIZE - i - 1]).toStrictEqual(tmpObj);
    }
});

test('shrink', () => {
    const shrinkThreshold = 30;
    pool = new Pool<Test>(createTest, 10, destroyTest, shrinkThreshold);
    calledTimes = 0;

    // Pool available size is 10, less than 30, so will not free any object.
    pool.tryShrink();
    // @ts-expect-error
    expect(pool._nextAvail).toEqual(9);
    calledTimes = 0;


    // Pool available size is 60 = 2 * shrinkThreshold, will shrink 30 elements. 
    const test: Test[] = [];
    for (let i = 0; i < 50; ++i) {
        test[i] = createTest();
    }
    pool.freeArray(test);
    pool.tryShrink();
    // @ts-expect-error
    expect(pool._nextAvail).toEqual(29);
    calledTimes = 30;

    // Pool available size is 50, but its length is 60, so it will shrink 10 elements.
    calledTimes = 0;
    for (let i = 0; i < 30; ++i) {
        pool.free(createTest());
    }
    for (let i = 0; i < 10; ++i) {
        pool.alloc();
    }
    pool.tryShrink();
    calledTimes = 10;
    // @ts-expect-error
    expect(pool._nextAvail).toEqual(39);

    // No matter shrink how many times, pool size will not less then shrinkThreshold.
    pool.tryShrink();
    pool.tryShrink();
    pool.tryShrink();
    pool.tryShrink();
    // @ts-expect-error
    expect(pool._nextAvail).toEqual(29);
});

test('destroy', () => {
    calledTimes = 0;
    // @ts-expect-error
    const beforeDestroy = pool._nextAvail;
    // @ts-expect-error
    expect(pool._nextAvail).toBeGreaterThan(0);

    pool.destroy();

    expect(calledTimes).toBe(beforeDestroy + 1);
    // @ts-expect-error
    expect(pool._nextAvail).toBe(-1);
    // @ts-expect-error
    expect(pool._freePool.length).toBe(0);
});

File: native/tests/unit-test/src/pool_test.cpp
/****************************************************************************
 Copyright (c) 2021 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/
#include "core/memop/Pool.h"
#include "gtest/gtest.h"

#include "utils.h"

using namespace cc;
using namespace memop;

namespace {
class PoolTest {
public:
    int32_t tag{0};
    explicit PoolTest(int32_t num) : tag(num) {}
};
int32_t tagValue = 0;

PoolTest *createTest() {
    return new PoolTest(tagValue++);
}

void destroyTest(PoolTest *obj) {
    delete obj;
}

const int32_t ARRAY_SIZE = 10;
auto *pool = new memop::Pool<PoolTest>(createTest, destroyTest, ARRAY_SIZE);

} // namespace

TEST(PoolTest, alloc) {
    for (int32_t i = 0; i < ARRAY_SIZE + 2; ++i) {
        auto *obj = pool->alloc();
        EXPECT_NE(obj, nullptr);
    }
}

TEST(PoolTest, free) {
    auto *tmp = createTest();
    pool->free(tmp);
    auto *newElement = pool->alloc();
    EXPECT_EQ(newElement, tmp);
}

TEST(PoolTest, freeArray) {
    const int32_t TEST_SIZE = 5;
    std::vector<PoolTest *> tmpArr(TEST_SIZE);
    for (int32_t i = 0; i < TEST_SIZE; ++i) {
        PoolTest *a = createTest();
    }
    pool->freeArray(tmpArr);
    for (int32_t i = 0; i < TEST_SIZE; ++i) {
        auto *tmpObj = pool->alloc();
        EXPECT_EQ(tmpArr[TEST_SIZE - i - 1], tmpObj);
    }
}
